Move "platforms" CI jobs into their own workflow files.

ScottTodd opened the pull request:
This moves workflow jobs that build the full project (including the compiler) from `ci.yml` to their own workflow files:

- `build_test_all_arm64` --> `.github/workflows/ci_linux_arm64_clang.yml`
- `build_test_all_windows` --> `.github/workflows/ci_windows_x86_msvc.yml`
- `build_test_all_macos_arm64` --> none
- `build_test_all_macos_x86_64` --> `.github/workflows/ci_macos_x86_clang.yml`

With one workflow file per platform, these are the tradeoffs compared to a single workflow file:

Pros:
+ Easy to tune which triggers affect each workflow
  * See https://github.com/Tencent/ncnn/blob/master/.github/workflows/windows-clang.yml for some inspiration. Each workflow can independently define `paths:` to trigger on. We have support for that via [`build_tools/github_actions/configure_ci.py`](https://github.com/iree-org/iree/blob/main/build_tools/github_actions/configure_ci.py), but this also allows for adjusting schedule triggers too.
+ Easy to fork a workflow for a new platform, less merge conflicts in a single large file
+ Easy to cancel a single job without affecting others from the same file
+ Easy to disable/enable a workflow in isolation from the UI if there are runner issues
+ Low impact radius if/when a single platform has runner issues

Cons:
- If running on PRs, checks are spread out a lot
- Hard to mark multiple/all jobs as required checks (if we care)
- More code duplication between files (reusable workflows can help here... maybe, could also lean on CMake presets and build scripts)

Practically at this point in time, these workflows _were_ running on all commits pushed to the `main` branch but they were disabled for being too slow or requiring runners that went offline. Now the workflows will run on a nightly schedule and they can be disabled individually from the GitHub UI as needed. TBD how we'll make these workflows available for opt-in on presubmit. We could add the `pull_request` trigger back, trigger if some labels are added, etc.

skip-ci: adding new workflows, existing ones are unaffected

ScottTodd committed:
Move macos CI job into its own workflow file.

ScottTodd committed:
Move Windows CI job into its own workflow.

ScottTodd committed:
Move Linux arm64 CI job into its own workflow.

benvanik commented on .github/workflows/ci_windows.yml:
```diff
@@ -0,0 +1,52 @@
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+name: CI - Windows
```
for consistency it'd be useful to have host arch in the CI names of everything (we're going to want windows_arm64 and then this is windows_x64 or whatever)

ScottTodd replied:
SG. Looking through https://github.com/Tencent/ncnn/tree/master/.github/workflows they also put the toolchain (when gcc vs clang).

ScottTodd replied:
> `build_test_all_macos_arm64` (no longer exists, this got confusing as we moved away from self-hosted mac runners)

https://docs.github.com/en/actions/using-jobs/choosing-the-runner-for-a-job#standard-github-hosted-runners-for-public-repositories
![image](https://github.com/user-attachments/assets/9c54509d-a4ca-4949-9c9e-1b40953a2b5f)


Am I reading this right?

`macos-13` is x86, while `macos-14` is arm (since these have M1 CPUs)

benvanik commented on .github/workflows/ci_linux_arm64.yml:
```diff
@@ -0,0 +1,60 @@
+jobs:
+  # TODO: Switch to GitHub-hosted arm64 runners when available?
+  # TODO: Switch runs-on labels to use different (non-GCP) self-hosted runners?
+  build_test_all_arm64:
```
can we rename this now to include platform for consistency? (build_test_all_linux_arm64, for whenever we have a build_test_all_<mac/windows/etc>)

ScottTodd replied:
Renamed. Also dropped the `build_test_all_` prefix.

ScottTodd commented on .github/workflows/ci_windows.yml:
```diff
@@ -0,0 +1,52 @@
+  cancel-in-progress: true
+
+jobs:
+  build_test_all_windows:
```
Sample logs from my fork: https://github.com/ScottTodd/iree/actions/runs/10034629151/job/27729434331

If we have enough of these nightly jobs taking 4h+ on GitHub-hosted runners then we may want to stagger them so they don't use up our limit of 20 concurrent actions (https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits)

(that had multiple jobs all grouped into a single "nightly" workflow)

ScottTodd committed:
Rename workflows to include platform/arch/compiler.

ScottTodd committed:
Adjust scheduled times to 01:15AM PST.

That should give these enough time to finish 4-5 hour runs before the US workday.

Timezones ugh...

ScottTodd committed:
Add back 'setup' job to arm64 workflow.

Could drop if changing cache usage or runner labels.

ScottTodd commented on .github/workflows/ci_linux_arm64_clang.yml:
```diff
@@ -0,0 +1,63 @@
+
+name: CI - Linux arm64 clang
+
+on:
+  schedule:
+    # Weekday mornings at 09:15 UTC = 01:15 PST (UTC - 8).
+    - cron: "15 9 * * 1-5"
+  workflow_dispatch:
```
@banach-space my thinking here is that you could go in to this file and change the triggers to fit your resourcing

For example: if you have a single runner (4+ cores) it could run nightly. If you have a single powerful runner (16+ cores), it could run on `push`. If you have multiple powerful runners (32+ cores), it could run on `pull_request`.

If runners go offline for whatever reason (cloud pool like GCP needs maintenance, self-hosted lab needs a local reboot, etc.), we can disable the job from the GitHub UI

---

I'm planning to sort of hack and slash through CI configuration this week and then follow up with some docs (maybe updating https://iree.dev/developers/general/contributing/#continuous-integration-ci) once the details have settled.

banach-space replied:
Thanks, this stuff is fairly straightforward, thanks for setting this up 🙏🏻

banach-space commented on .github/workflows/ci_macos_x86_clang.yml:
```diff
@@ -0,0 +1,58 @@
+
+jobs:
+  macos_x86_clang:
+    runs-on: macos-13
```
@ScottTodd Have you considered using M1 runners? https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/ :)

ScottTodd replied:
Yeah. We're testing the runtime on macos-14 (m1) right now. Not sure if I've tested building the compiler on those runners too. Should be easy enough to add a new build configuration.

banach-space commented on .github/workflows/ci_linux_arm64_clang.yml:
```diff
@@ -0,0 +1,63 @@
+          IREE_WRITE_REMOTE_CCACHE: ${{ needs.setup.outputs.write-caches }}
+        run: |
+          ./build_tools/github_actions/docker_run.sh \
+            --env "IREE_CCACHE_GCP_TOKEN=$(gcloud auth application-default print-access-token)" \
```
I am guessing that this is setting up some CCACHE magic specific to GCP? Something that wouldn't work on self-hosted runners outside of GCP?

ScottTodd replied:
Yep. I just copied the previous job to this new file with only minor edits. The other jobs aren't using any caching at the moment, so they are pretty slow. If you switch this to your own runner, you might want to remove/adjust the Docker code and use a different (or no) cache mechanism.

We've experimented with different caching options:
* Self-hosted persistent runners with local ccache
* Self-hosted ephemeral runners with cloud ccache (e.g. GCP/GCS)
* GitHub-hosted runners with actions caches: https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy

Local caches on persistent runners are simplest :)

Cloud caches near the build machines (e.g. GCP using GCS) should have good network latency and bandwidth, as well as growable sizes and configurable TTLs / eviction policies.

GitHub's caches are pretty limited - 10GB across all caches in the repository before old entries start getting evicted. I remember 2-3GB caches being slow to upload/download, and the compiler typically needs ~3GB to get the 95%+ cache hit rates that speed up the build enough. In my experiments, if the cache hit rate got to ~80% then the build time would jump up substantially (it's been a while, but IIRC it was like 15 minutes --> 40 minutes).

banach-space replied:
> 95%+ cache hit rates 

Presumably for a clean build? I've never hit such numbers for incremental builds. This is beside the point, I'm just curious :)

ScottTodd replied:
Hmmm... Not sure for arm64 with clang. Many of my experiments were for Windows with MSVC.

arm64 from last night:
* 1h30m first build (cold cache): https://github.com/iree-org/iree/actions/runs/10049822259/job/27776581816#step:4:10448
    ```
    + ccache --show-stats
    Cacheable calls:   5709 / 5729 (99.65%)
      Hits:             660 / 5709 (11.56%)
        Direct:         144 /  660 (21.82%)
        Preprocessed:   516 /  660 (78.18%)
      Misses:          5049 / 5709 (88.44%)
    Uncacheable calls:   20 / 5729 ( 0.35%)
    Local storage:
      Cache size (GB): 0.00 / 5.00 ( 0.00%)
      Hits:               0
      Misses:             0
    Remote storage:
      Hits:             660 / 5716 (11.55%)
      Misses:          5056 / 5716 (88.45%)
      Errors:          5233
    ```
* 50m second build (warm cache): https://github.com/iree-org/iree/actions/runs/10056339628/job/27794936386#step:4:10453
    ```
    + ccache --show-stats
    Cacheable calls:   5709 / 5729 (99.65%)
      Hits:            3428 / 5709 (60.05%)
        Direct:         931 / 3428 (27.16%)
        Preprocessed:  2497 / 3428 (72.84%)
      Misses:          2281 / 5709 (39.95%)
    Uncacheable calls:   20 / 5729 ( 0.35%)
    Local storage:
      Cache size (GB): 0.00 / 5.00 ( 0.00%)
      Hits:               0
      Misses:             0
    Remote storage:
      Hits:            3428 / 5715 (59.98%)
      Misses:          2287 / 5715 (40.02%)
      Errors:          4379
    ```


Compare that with linux x86 clang (on a VM with more cores):
* 13m build after an LLVM integrate (so cold cache): https://github.com/iree-org/iree/actions/runs/10012178032/job/27677127470#step:4:10933
    ```
    + ccache --show-stats
    Cacheable calls:   5802 / 5822 (99.66%)
      Hits:            2335 / 5802 (40.24%)
        Direct:         824 / 2335 (35.29%)
        Preprocessed:  1511 / 2335 (64.71%)
      Misses:          3467 / 5802 (59.76%)
    Uncacheable calls:   20 / 5822 ( 0.34%)
    Local storage:
      Cache size (GB): 0.00 / 5.00 ( 0.00%)
      Hits:               0
      Misses:             0
    Remote storage:
      Hits:            2335 / 5810 (40.19%)
      Misses:          3475 / 5810 (59.81%)
      Errors:          4497
    ```
* 5m build after a no-op (so warm cache): https://github.com/iree-org/iree/actions/runs/10012585025/job/27678397517#step:4:10933
    ```
    + ccache --show-stats
    Cacheable calls:   5802 / 5822 (99.66%)
      Hits:            5801 / 5802 (99.98%)
        Direct:        1006 / 5801 (17.34%)
        Preprocessed:  4795 / 5801 (82.66%)
      Misses:             1 / 5802 ( 0.02%)
    Uncacheable calls:   20 / 5822 ( 0.34%)
    Local storage:
      Cache size (GB): 0.00 / 5.00 ( 0.00%)
      Hits:               0
      Misses:             0
    Remote storage:
      Hits:            5801 / 5810 (99.85%)
      Misses:             9 / 5810 ( 0.15%)
      Errors:          4439
    ```
